// @remotion/renderer/dist/call-ffmpeg.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.callFf = exports.callFfExtraOptions = void 0;
const execa_1 = __importDefault(require("execa"));
const path_1 = __importDefault(require("path"));
const get_executable_path_1 = require("./compositor/get-executable-path");
const truthy_1 = require("./truthy");
const callFfExtraOptions = () => {
    const lib = path_1.default.join((0, get_executable_path_1.getExecutablePath)('ffmpeg-cwd'), 'remotion', 'lib');
    return {
        env: process.platform === 'darwin'
            ? {
                DYLD_LIBRARY_PATH: lib,
            }
            : process.platform === 'win32'
                ? {
                    PATH: `${process.env.PATH};${lib}`,
                }
                : {
                    LD_LIBRARY_PATH: lib,
                },
    };
};
exports.callFfExtraOptions = callFfExtraOptions;
const callFf = (bin, args, options) => {
    return (0, execa_1.default)((0, get_executable_path_1.getExecutablePath)(bin), args.filter(truthy_1.truthy), {
        ...(0, exports.callFfExtraOptions)(),
        ...options,
    });
};
exports.callFf = callFf;
