// @remotion/renderer/dist/codec-supports-media.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.codecSupportsCrf = exports.codecSupportsMedia = void 0;
const get_codec_name_1 = require("./get-codec-name");
const support = {
    'h264-mkv': {
        audio: true,
        video: true,
    },
    aac: {
        audio: true,
        video: false,
    },
    gif: {
        video: true,
        audio: false,
    },
    h264: {
        video: true,
        audio: true,
    },
    h265: {
        video: true,
        audio: true,
    },
    mp3: {
        audio: true,
        video: false,
    },
    prores: {
        audio: true,
        video: true,
    },
    vp8: {
        audio: true,
        video: true,
    },
    vp9: {
        audio: true,
        video: true,
    },
    wav: {
        audio: true,
        video: false,
    },
};
const codecSupportsMedia = (codec) => {
    return support[codec];
};
exports.codecSupportsMedia = codecSupportsMedia;
const codecSupportsCrf = (codec) => {
    const encoderName = (0, get_codec_name_1.getCodecName)(codec);
    const supportsCrf = encoderName && codec !== 'prores';
    return supportsCrf;
};
exports.codecSupportsCrf = codecSupportsCrf;
