// @remotion/renderer/dist/file-extensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultFileExtensionMap = exports.supportedAudioCodecs = void 0;
exports.supportedAudioCodecs = {
    h264: ['aac', 'pcm-16'],
    'h264-mkv': ['pcm-16'],
    aac: ['aac', 'pcm-16'],
    gif: [],
    h265: ['aac', 'pcm-16'],
    mp3: ['mp3', 'pcm-16'],
    prores: ['pcm-16', 'aac'],
    vp8: ['opus', 'pcm-16'],
    vp9: ['opus', 'pcm-16'],
    wav: ['pcm-16'],
};
exports.defaultFileExtensionMap = {
    'h264-mkv': {
        default: 'mkv',
        forAudioCodec: {
            'pcm-16': { possible: ['mkv'], default: 'mkv' },
        },
    },
    aac: {
        default: 'aac',
        forAudioCodec: {
            aac: {
                possible: ['aac', '3gp', 'm4a', 'm4b', 'mpg', 'mpeg'],
                default: 'aac',
            },
            'pcm-16': {
                possible: ['wav'],
                default: 'wav',
            },
        },
    },
    gif: {
        default: 'gif',
        forAudioCodec: {},
    },
    h264: {
        default: 'mp4',
        forAudioCodec: {
            'pcm-16': { possible: ['mkv', 'mov'], default: 'mkv' },
            aac: { possible: ['mp4', 'mkv', 'mov'], default: 'mp4' },
        },
    },
    h265: {
        default: 'mp4',
        forAudioCodec: {
            aac: { possible: ['mp4', 'mkv', 'hevc'], default: 'mp4' },
            'pcm-16': { possible: ['mkv'], default: 'mkv' },
        },
    },
    mp3: {
        default: 'mp3',
        forAudioCodec: {
            mp3: { possible: ['mp3'], default: 'mp3' },
            'pcm-16': { possible: ['wav'], default: 'wav' },
        },
    },
    prores: {
        default: 'mov',
        forAudioCodec: {
            aac: { possible: ['mov', 'mkv', 'mxf'], default: 'mov' },
            'pcm-16': { possible: ['mov', 'mkv', 'mxf'], default: 'mov' },
        },
    },
    vp8: {
        default: 'webm',
        forAudioCodec: {
            'pcm-16': { possible: ['mkv'], default: 'mkv' },
            opus: { possible: ['webm'], default: 'webm' },
        },
    },
    vp9: {
        default: 'webm',
        forAudioCodec: {
            'pcm-16': { possible: ['mkv'], default: 'mkv' },
            opus: { possible: ['webm'], default: 'webm' },
        },
    },
    wav: {
        default: 'wav',
        forAudioCodec: {
            'pcm-16': { possible: ['wav'], default: 'wav' },
        },
    },
};
