// @remotion/cli/dist/editor/components/Timeline/imperative-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setCurrentFps = exports.getCurrentFps = exports.setCurrentDuration = exports.getCurrentDuration = exports.setCurrentFrame = exports.getCurrentFrame = exports.setCurrentZoom = exports.getCurrentZoom = exports.currentFps = exports.currentDuration = exports.currentZoom = exports.currentFrame = void 0;
exports.currentFrame = 0;
exports.currentZoom = 1;
exports.currentDuration = 1;
exports.currentFps = 1;
const getCurrentZoom = () => {
    return exports.currentZoom;
};
exports.getCurrentZoom = getCurrentZoom;
const setCurrentZoom = (z) => {
    exports.currentZoom = z;
};
exports.setCurrentZoom = setCurrentZoom;
const getCurrentFrame = () => {
    return exports.currentFrame;
};
exports.getCurrentFrame = getCurrentFrame;
const setCurrentFrame = (f) => {
    exports.currentFrame = f;
};
exports.setCurrentFrame = setCurrentFrame;
const getCurrentDuration = () => {
    return exports.currentDuration;
};
exports.getCurrentDuration = getCurrentDuration;
const setCurrentDuration = (d) => {
    exports.currentDuration = d;
};
exports.setCurrentDuration = setCurrentDuration;
const getCurrentFps = () => {
    return exports.currentFps;
};
exports.getCurrentFps = getCurrentFps;
const setCurrentFps = (d) => {
    exports.currentFps = d;
};
exports.setCurrentFps = setCurrentFps;
