// @remotion/cli/dist/editor/components/CollapsedSidebarExpander.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CollapsedSidebarExpander = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const colors_1 = require("../helpers/colors");
const caret_1 = require("../icons/caret");
const z_index_1 = require("../state/z-index");
// TODO: Add shortcut to toggle both sidebars (similar to zen mode in vs code)
const CollapsedSidebarExpander = ({ onExpand, direction }) => {
    const [hovered, setHovered] = (0, react_1.useState)(false);
    const { tabIndex } = (0, z_index_1.useZIndex)();
    const onPointerEnter = (0, react_1.useCallback)(() => {
        setHovered(true);
    }, []);
    const onPointerLeave = (0, react_1.useCallback)(() => {
        setHovered(false);
    }, []);
    const style = (0, react_1.useMemo)(() => {
        return {
            border: 'none',
            borderRight: `${direction === 'right' ? 1 : 0}px solid black`,
            borderLeft: `${direction === 'left' ? 1 : 0}px solid black`,
            cursor: 'pointer',
            color: 'white',
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
            paddingLeft: direction === 'right' ? 7 : 5,
            paddingRight: direction === 'right' ? 4 : 5,
            backgroundColor: (0, colors_1.getBackgroundFromHoverState)({
                hovered,
                selected: false,
            }),
            appearance: 'none',
            WebkitAppearance: 'none',
        };
    }, [direction, hovered]);
    return ((0, jsx_runtime_1.jsx)("button", { style: style, onPointerEnter: onPointerEnter, onPointerLeave: onPointerLeave, type: "button", role: "button", tabIndex: tabIndex, onClick: onExpand, children: direction === 'right' ? (0, jsx_runtime_1.jsx)(caret_1.CaretRight, {}) : (0, jsx_runtime_1.jsx)(caret_1.CaretLeft, {}) }));
};
exports.CollapsedSidebarExpander = CollapsedSidebarExpander;
