// @remotion/cli/dist/editor/components/InlineAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InlineAction = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const colors_1 = require("../helpers/colors");
const z_index_1 = require("../state/z-index");
const InlineAction = ({ children, onClick }) => {
    const { tabIndex } = (0, z_index_1.useZIndex)();
    const [hovered, setHovered] = (0, react_1.useState)(false);
    const onPointerEnter = (0, react_1.useCallback)(() => {
        setHovered(true);
    }, []);
    const onPointerLeave = (0, react_1.useCallback)(() => {
        setHovered(false);
    }, []);
    const style = (0, react_1.useMemo)(() => {
        return {
            border: 'none',
            background: hovered ? colors_1.CLEAR_HOVER : 'transparent',
            height: 24,
            width: 24,
            display: 'inline-flex',
            justifyContent: 'center',
            alignItems: 'center',
            borderRadius: 3,
            // TODO: Color does not get propagated to the children
            color: hovered ? 'white' : colors_1.LIGHT_TEXT,
        };
    }, [hovered]);
    return (
    // <div> because cannot use button inside a button
    (0, jsx_runtime_1.jsx)("a", { onPointerEnter: onPointerEnter, onPointerLeave: onPointerLeave, onClick: onClick, style: style, tabIndex: tabIndex, children: children }));
};
exports.InlineAction = InlineAction;
