// @remotion/cli/dist/select-composition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.selectCompositions = exports.selectComposition = void 0;
const composition_prompts_1 = require("./composition-prompts");
const chalk_1 = require("./chalk");
const selectComposition = async (validCompositions) => {
    if (validCompositions.length === 1) {
        const onlyComposition = validCompositions[0];
        if (onlyComposition) {
            return {
                compositionId: onlyComposition.id,
                reason: 'Only composition',
            };
        }
    }
    const selectedComposition = (await (0, composition_prompts_1.selectAsync)({
        message: 'Select composition:',
        optionsPerPage: 5,
        type: 'select',
        choices: validCompositions.map((comp) => {
            return {
                value: comp.id,
                title: chalk_1.chalk.bold(comp.id),
            };
        }),
    }));
    return { compositionId: selectedComposition, reason: 'Selected' };
};
exports.selectComposition = selectComposition;
const selectCompositions = async (validCompositions) => {
    if (validCompositions.length === 1) {
        const onlyComposition = validCompositions[0];
        if (onlyComposition) {
            return [onlyComposition.id];
        }
    }
    const selectedComposition = await (0, composition_prompts_1.selectAsync)({
        message: 'Select compositions:',
        optionsPerPage: 5,
        type: 'multiselect',
        min: 1,
        choices: validCompositions.map((comp) => {
            return {
                value: comp.id,
                title: chalk_1.chalk.bold(comp.id),
            };
        }),
    });
    return selectedComposition;
};
exports.selectCompositions = selectCompositions;
