// @remotion/cli/dist/setup-cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bundleOnCli = exports.bundleOnCliOrTakeServeUrl = void 0;
const bundler_1 = require("@remotion/bundler");
const renderer_1 = require("@remotion/renderer");
const config_1 = require("./config");
const log_1 = require("./log");
const parse_command_line_1 = require("./parse-command-line");
const progress_bar_1 = require("./progress-bar");
const bundleOnCliOrTakeServeUrl = async ({ fullPath, remotionRoot, steps, publicDir, }) => {
    if (renderer_1.RenderInternals.isServeUrl(fullPath)) {
        return {
            urlOrBundle: fullPath,
            cleanup: () => Promise.resolve(undefined),
        };
    }
    const bundled = await (0, exports.bundleOnCli)({ fullPath, remotionRoot, steps, publicDir });
    return {
        urlOrBundle: bundled,
        cleanup: () => renderer_1.RenderInternals.deleteDirectory(bundled),
    };
};
exports.bundleOnCliOrTakeServeUrl = bundleOnCliOrTakeServeUrl;
const bundleOnCli = async ({ fullPath, steps, remotionRoot, publicDir, }) => {
    var _a;
    const shouldCache = config_1.ConfigInternals.getWebpackCaching();
    const symlinkState = {
        symlinks: [],
    };
    const onProgress = (progress) => {
        bundlingState = {
            progress: progress / 100,
            steps,
            doneIn: null,
        };
        bundlingProgress.update((0, progress_bar_1.makeBundlingAndCopyProgress)({
            bundling: bundlingState,
            copying: copyingState,
            symLinks: symlinkState,
        }));
    };
    let copyingState = {
        bytes: 0,
        doneIn: null,
    };
    let copyStart = null;
    const updateProgress = (newline) => {
        bundlingProgress.update((0, progress_bar_1.makeBundlingAndCopyProgress)({
            bundling: bundlingState,
            copying: copyingState,
            symLinks: symlinkState,
        }) + (newline ? '\n' : ''));
    };
    const onPublicDirCopyProgress = (bytes) => {
        if (copyStart === null) {
            copyStart = Date.now();
        }
        copyingState = {
            bytes,
            doneIn: null,
        };
        updateProgress(false);
    };
    const onSymlinkDetected = (absPath) => {
        symlinkState.symlinks.push(absPath);
        updateProgress(false);
    };
    const options = {
        enableCaching: shouldCache,
        webpackOverride: (_a = config_1.ConfigInternals.getWebpackOverrideFn()) !== null && _a !== void 0 ? _a : ((f) => f),
        rootDir: remotionRoot,
        publicDir,
        onPublicDirCopyProgress,
        onSymlinkDetected,
    };
    const [hash] = bundler_1.BundlerInternals.getConfig({
        outDir: '',
        entryPoint: fullPath,
        onProgress,
        options,
        resolvedRemotionRoot: remotionRoot,
    });
    const cacheExistedBefore = bundler_1.BundlerInternals.cacheExists(remotionRoot, 'production', hash);
    if (cacheExistedBefore !== 'does-not-exist' && !shouldCache) {
        log_1.Log.info('🧹 Cache disabled but found. Deleting... ');
        await bundler_1.BundlerInternals.clearCache(remotionRoot);
    }
    if (cacheExistedBefore === 'other-exists' && shouldCache) {
        log_1.Log.info('🧹 Webpack config change detected. Clearing cache... ');
        await bundler_1.BundlerInternals.clearCache(remotionRoot);
    }
    const bundleStartTime = Date.now();
    const bundlingProgress = (0, progress_bar_1.createOverwriteableCliOutput)((0, parse_command_line_1.quietFlagProvided)());
    let bundlingState = {
        progress: 0,
        steps,
        doneIn: null,
    };
    const bundled = await (0, bundler_1.bundle)({
        entryPoint: fullPath,
        onProgress: (progress) => {
            bundlingState = {
                progress: progress / 100,
                steps,
                doneIn: null,
            };
            updateProgress(false);
        },
        ...options,
    });
    bundlingState = {
        progress: 1,
        steps,
        doneIn: Date.now() - bundleStartTime,
    };
    copyingState = {
        ...copyingState,
        doneIn: copyStart ? Date.now() - copyStart : null,
    };
    updateProgress(true);
    log_1.Log.verbose('Bundled under', bundled);
    const cacheExistedAfter = bundler_1.BundlerInternals.cacheExists(remotionRoot, 'production', hash) === 'exists';
    if (cacheExistedAfter) {
        if (cacheExistedBefore === 'does-not-exist' ||
            cacheExistedBefore === 'other-exists') {
            log_1.Log.info('⚡️ Cached bundle. Subsequent renders will be faster.');
        }
    }
    return bundled;
};
exports.bundleOnCli = bundleOnCli;
