// @remotion/cli/dist/editor/components/RenderQueue/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateDefaultProps = exports.cancelRenderJob = exports.removeRenderJob = exports.openInFileExplorer = exports.subscribeToFileExistenceWatcher = exports.unsubscribeFromFileExistenceWatcher = exports.addVideoRenderJob = exports.addStillRenderJob = exports.callApi = void 0;
const date_serialization_1 = require("../RenderModal/SchemaEditor/date-serialization");
const callApi = (endpoint, body) => {
    return new Promise((resolve, reject) => {
        fetch(endpoint, {
            method: 'post',
            headers: {
                'content-type': 'application/json',
            },
            body: JSON.stringify(body),
        })
            .then((res) => res.json())
            .then((data) => {
            if (data.success) {
                resolve(data.data);
            }
            else {
                reject(new Error(data.error));
            }
        })
            .catch((err) => {
            reject(err);
        });
    });
};
exports.callApi = callApi;
const addStillRenderJob = ({ compositionId, outName, imageFormat, quality, frame, scale, verbose, chromiumOptions, delayRenderTimeout, envVariables, inputProps, }) => {
    return (0, exports.callApi)('/api/render', {
        compositionId,
        type: 'still',
        outName,
        imageFormat,
        quality,
        frame,
        scale,
        verbose,
        chromiumOptions,
        delayRenderTimeout,
        envVariables,
        inputProps,
    });
};
exports.addStillRenderJob = addStillRenderJob;
const addVideoRenderJob = ({ compositionId, outName, imageFormat, quality, scale, verbose, codec, concurrency, crf, startFrame, endFrame, muted, enforceAudioTrack, proResProfile, pixelFormat, audioBitrate, videoBitrate, everyNthFrame, numberOfGifLoops, delayRenderTimeout, audioCodec, disallowParallelEncoding, chromiumOptions, envVariables, inputProps, }) => {
    return (0, exports.callApi)('/api/render', {
        compositionId,
        type: 'video',
        outName,
        imageFormat,
        quality,
        scale,
        verbose,
        codec,
        concurrency,
        crf,
        endFrame,
        startFrame,
        muted,
        enforceAudioTrack,
        proResProfile,
        pixelFormat,
        audioBitrate,
        videoBitrate,
        everyNthFrame,
        numberOfGifLoops,
        delayRenderTimeout,
        audioCodec,
        disallowParallelEncoding,
        chromiumOptions,
        envVariables,
        inputProps,
    });
};
exports.addVideoRenderJob = addVideoRenderJob;
const unsubscribeFromFileExistenceWatcher = ({ file, clientId, }) => {
    return (0, exports.callApi)('/api/unsubscribe-from-file-existence', { file, clientId });
};
exports.unsubscribeFromFileExistenceWatcher = unsubscribeFromFileExistenceWatcher;
const subscribeToFileExistenceWatcher = async ({ file, clientId, }) => {
    const { exists } = await (0, exports.callApi)('/api/subscribe-to-file-existence', {
        file,
        clientId,
    });
    return exists;
};
exports.subscribeToFileExistenceWatcher = subscribeToFileExistenceWatcher;
const openInFileExplorer = ({ directory }) => {
    const body = {
        directory,
    };
    return (0, exports.callApi)('/api/open-in-file-explorer', body);
};
exports.openInFileExplorer = openInFileExplorer;
const removeRenderJob = (job) => {
    return (0, exports.callApi)('/api/remove-render', {
        jobId: job.id,
    });
};
exports.removeRenderJob = removeRenderJob;
const cancelRenderJob = (job) => {
    return (0, exports.callApi)('/api/cancel', {
        jobId: job.id,
    });
};
exports.cancelRenderJob = cancelRenderJob;
const updateDefaultProps = (compositionId, defaultProps) => {
    return (0, exports.callApi)('/api/update-default-props', {
        compositionId,
        defaultProps: (0, date_serialization_1.serializeJSONWithDate)(defaultProps, undefined),
    });
};
exports.updateDefaultProps = updateDefaultProps;
