// @remotion/cli/dist/preview-server/error-overlay/remotion-overlay/Button.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Button = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const colors_1 = require("../../../editor/helpers/colors");
const button = {
    border: `1px solid ${colors_1.INPUT_BORDER_COLOR_UNHOVERED}`,
    borderRadius: 4,
    backgroundColor: colors_1.INPUT_BACKGROUND,
    appearance: 'none',
    fontFamily: 'inherit',
    fontSize: 14,
    color: 'white',
    flexDirection: 'row',
};
const buttonContainer = {
    padding: 10,
    cursor: 'pointer',
    fontSize: 14,
};
const Button = ({ children, onClick, disabled, style }) => {
    const combined = (0, react_1.useMemo)(() => {
        return {
            ...button,
            ...(style !== null && style !== void 0 ? style : {}),
        };
    }, [style]);
    return ((0, jsx_runtime_1.jsx)("button", { style: combined, type: "button", disabled: disabled, onClick: onClick, children: (0, jsx_runtime_1.jsx)("div", { className: "css-reset", style: buttonContainer, children: children }) }));
};
exports.Button = Button;
